### ISSUE TITLE
when karma hangs, how do I debug it?

### ISSUE THREAD

**Issue by cellog** [karma-runner/karma#2211]
In my project, https://github.com/cellog/react-selection , I have encountered a bug when running karma remotely on saucelabs.  It is caused by code coverage.  In short, karma connects to the browser, loads, and then hangs indefinitely.  When I remove code coverage, the test runs in about 16 seconds.  This happens only remotely, when I run coverage on my local machine, it works great.

How can I debug the hang on saucelabs?  I'm new to karma.  When using karma remotely with saucelabs, is it loading karma from my machine?  If yes, which file can I put debug code into to see if I can figure out where it is failing?  Or is there another way?

Thanks much.

**Comment by cellog** [comment]
Oh and I should mention the code coverage is accomplished by a babel plugin, so there is nothing inside karma, it just serving the transpiled files directly to saucelabs

**Comment by dignifiedquire** [comment]
setting `logLevel` to `DEBUG` should give you more details, after that regular debugging techniques for node loke console.log in code and attaching a debugger

**Comment by cellog** [comment]
it turns out that any concurrency larger than 1 causes the entire thing to fail with disconnects to every browser.  Is this expected behavior?

**Comment by dignifiedquire** [comment]
On saucelabs that is probably caused by rate limiting from their side.

**Comment by cellog** [comment]
fyi, the problem is in the karma-saucelabs-launcher, which is unmaintained and doesn't work any more.  After 50 tests, it just dies.  I switched to browserstack and it works perfectly.  However, there are some side effects of running more than 1 browser test in the same karma process.  I have no idea how to isolate them, but the ONLY way I can get karma to work *and* generate code coverage is to run each browser in its own karma process.  You can see what I mean in https://github.com/cellog/react-selection

Major pain.

Fortunately, I can do local development with karma very quickly, and then push on commit to find those chance browser differences eventually.  Test runs now take about 15 minutes each, so it's really a pain, but at least it works.

So to summarize: there are several bugs that I can't fix or even track down.  The first is that karma and sauce labs are no longer friends, and so I can't use them together at all.  The second is that karma has some kind of strange shared stuff interfering with each other whenever I run more than 1 browser in the same karma process, even if concurrency is set to 1.  That's a huge one, and might be worth investigating further, since it means the sandboxing is leaking.  Let me know if you want me to try things to debug it.

**Comment by wesleycho** [comment]
FWIW, I have noticed that the Angular team has set up karma with Sauce Labs to use Sauce Labs via shell script in separate processes - if I had to guess, it is to have environmental separation for purer test environments, but maybe they came across some of the same issues with multiple browsers.

Separate issues should probably be opened if there are specific issues found though, that way we can track them better. Going to close this issue, but feel free to open issues for actionable problems you find with karma.

**Comment by trusktr** [comment]
> attaching a debugger

How exactly? I tried

`ndb ./node_modules/.bin/karma ...` but nothing runs, and the opened devtools window is empty, no source files, nothing paused.

**Comment by cellog** [comment]
The correct answer is to move to jest and forget about it

**Comment by johnjbarton** [comment]
Unfortunately I've found ndb fail sometimes. I use `node  --inspect-brk` but I also don't use the .bin file directly.
